doc/Messaging.md

```diff
@@ -0,0 +1,140 @@
+
+Although messages will generally be delivered in the order they're created, this is not an explicit guarantee of the system. The focus is more on keeping track of the work that needs to be done and ensuring that it was performed. Messages are good for:
+
+* Handing off work to another system.
+* Recording potentially time-consuming work that needs to be done asynchronously.
+* Scheduling for future delivery.
+* Accumulating work that could be done during off-peak hours.
+
+Messages are not a good fit for the following use cases:
+
+* Broadcasting of events to multiple subscribers.
+* Ordered delivery.
+* Real-time delivery.
+
+# Creating a message table
+
+The current implementation requires a fixed schema. This will be made more flexible in the future. There will also be a custom DDL syntax. For now, a message table must be created like this:
+
+```
+create table my_message(
+  time_scheduled bigint,
+  id bigint,
+  time_next bigint,
+  epoch bigint,
+  time_created bigint,
+  time_acked bigint,
+  message varchar(128),
+  primary key(time_scheduled, id),
+  unique index id_idx(id),
+  index next_idx(time_next, epoch)
+) comment 'vitess_message,vt_ack_wait=30,vt_purge_after=86400,vt_batch_size=10,vt_cache_size=10000,vt_poller_interval=30'
+```
+
+The application-related columns are as follows:
+
+* `id`: can be any type. Must be unique.
+* `message`: can be any type.
+* `time_scheduled`: must be a bigint. It will be used to store unix time in nanoseconds. If unspecified, the `Now` value is inserted.
+
+The above indexes are recommended for optimum performance. However, some variation can be allowed to achieve different performance trade-offs.
+
+The comment section specifies additional configuration parameters. The fields are as follows:
+
+* `vitess_message`: Indicates that this is a message table.
+* `vt_ack_wait=30`: Wait for 30s for the first message ack. If one is not received, resend.
+* `vt_purge_after=86400`: Purge acked messages that are older than 86400 seconds (1 day).
+* `vt_batch_size=10`: Send up to 10 messages per batch.
+* `vt_cache_seze=10000`: Store up to 10000 messages in the cache.
+* `vt_poller_interval=30`: Poll every 30s for messages that are due to be sent.
+
+If any of the above fields are missing, vitess will fail to load the table. No operation will be allowed on a table that has failed to load.
+
+# Enqueuing messages
+
+The application can enqueue messages using an insert statement:
+
+```
+insert into my_message(id, message) values(1, 'hello world')
+```
+
+These inserts can be part of a regular transaction. Multiple messages can be inserted to different tables. Avoid accumulating too many big messages within a transaction as it consumes memory on the VTTablet side. At the time of commit, memoery permitting, all messages are instantly enqueued to be sent.
```

Comment:
What does that VTTablet memory limitation look like? It's pretty common for us to add > 1M tasks in a single transaction, with the payload being a ~100 character url.